source/common/http/utility.cc

```diff
@@ -748,48 +751,153 @@ std::string Utility::PercentEncoding::decode(absl::string_view encoded) {
     }
   }
 
-  // Now see if this is an IP address. We need to know this because some things (such as setting
-  // SNI) are special cased if this is an IP address. Either way, we still go through the normal
-  // resolver flow. We could short-circuit the DNS resolver in this case, but the extra code to do
-  // so is not worth it since the DNS resolver should handle it for us.
-  bool is_ip_address = false;
+  // index number of colon is less than 6
+  if (port_pos != -1) {
+    return authority.substr(0, port_pos);
+  }
+
+  const auto port_flag_pos = rightLimitedFind(authority, ':', 6);
+  if (port_flag_pos != absl::nullopt) {
+    return authority.substr(0, port_flag_pos.value());
+  }
+
+  return authority;
+}
+
+absl::optional<absl::string_view> Utility::portFromAuthority(absl::string_view authority,
+                                                             int port_pos) {
+  if (port_pos != -1) {
+    return authority.substr(port_pos + 1);
+  }
+
+  const auto pos = rightLimitedFind(authority, ':', 6);
+  if (pos == absl::nullopt) {
+    return absl::nullopt;
+  }
+
+  // port section may be not include more than 5 digits, so that should be treated as malformed
+  if (authority.substr(pos.value() + 1).size() > 5) {
+    throw EnvoyException(fmt::format("malformed authority: {}", authority));
+  }
+  return authority.substr(pos.value() + 1);
+}
+
+void Utility::AttributeUpdater(AuthorityAttributes& auth_attr, absl::string_view authority,
+                               int port_pos) {
   try {
-    absl::string_view potential_ip_address = host_to_resolve;
-    // TODO(mattklein123): Optimally we would support bracket parsing in parseInternetAddress(),
-    // but we still need to trim the brackets to send the IPv6 address into the DNS resolver. For
-    // now, just do all the trimming here, but in the future we should consider whether we can
-    // have unified [] handling as low as possible in the stack.
-    if (potential_ip_address.front() == '[' && potential_ip_address.back() == ']') {
-      potential_ip_address.remove_prefix(1);
-      potential_ip_address.remove_suffix(1);
+    if (port_pos != -1) {
+      const Network::Address::InstanceConstSharedPtr instance =
+          Network::Utility::parseInternetAddressAndPort(authority);
+      auth_attr.host = Http::Utility::hostFromAuthority(authority, port_pos);
+      auth_attr.port = instance->ip()->port();
+      auth_attr.is_ip_address = true;
+    } else {
+      auto extracted_host = Http::Utility::hostFromAuthority(authority);
+      Network::Utility::parseInternetAddress(extracted_host);
+      auth_attr.host = extracted_host;
     }
-    Network::Utility::parseInternetAddress(std::string(potential_ip_address));
-    is_ip_address = true;
-    host_to_resolve = potential_ip_address;
+    auth_attr.is_ip_address = true;
   } catch (const EnvoyException&) {
   }
+}
+
+const Utility::AuthorityAttributes Utility::parseAuthority(absl::string_view authority) {
```

Comment:
Sorry for the delays. After spending more time looking at the changes it is not clear to me what exactly this PR is changing. The previous algorithm was the same first-match-wins and returns the same values as the new algorithm. The new algorithm is doing more parsing, but I do not think it creates any difference in the visible behavior of the parseAuthority method. Unit tests have not changed.
To clarify these issues, could you add unit tests that fail with the old algorithm and pass with the new changes, please?

Reply:
@yanavlasov I think this is only refactoring of current authority parsing. So, there is no changes about test pattern. and there is string to string_view changes. This may help performance improvement. But, it may seem that almost all of sections are useless. If this addition may cause vulnerable or more performance suffering, This PR should be disposed.

Reply:
I think I've got a bit lost with so many commits in this PR. This refactor adds quite a bit more code and more checks (i.e. isIpv4Address) however the reason for these additional checks is not expressed through unit tests.
I'm also uncertain about performance improvements are you have replaced one string_view->string conversion with two.
In this final form, I'm unsure this PR is justified.